### ISSUE TITLE
storage: too many open files in addSSTablePreApply on repeated create,restore,drop iteration

### ISSUE THREAD

**Issue by robert-s-lee** [cockroachdb/cockroach#17817]
**BUG REPORT**

1. Please supply the header (i.e. the first few lines) of your most recent
   log file **for each node in your cluster**. On most unix-based systems

 $ cockroach version
Build Tag:    v1.1-alpha.20170817
Build Time:   2017/08/17 22:27:51
Distribution: CCL
Platform:     darwin amd64
Go Version:   go1.8.3
C Compiler:   4.2.1 Compatible Clang 3.8.0 (tags/RELEASE_380/final)
Build SHA-1:  9a357a9f553a21ec08235492c7b916219652581a
Build Type:   release

2. Please describe the issue you observed:

- What did you do?

root@:26257/> create database tpch;
CREATE DATABASE

Time: 14.083753ms

root@:26257/> restore tpch.* FROM 'nodelocal:/Users/rslee/data/backup/tpch';
+--------------------+-----------+--------------------+---------+---------------+----------------+------------+
|       job_id       |  status   | fraction_completed |  rows   | index_entries | system_records |   bytes    |
+--------------------+-----------+--------------------+---------+---------------+----------------+------------+
| 273132076805947392 | succeeded |                  1 | 8661245 |      56229775 |              0 | 1866390204 |
+--------------------+-----------+--------------------+---------+---------------+----------------+------------+
(1 row)

Time: 59.59842683s

root@:26257/> drop database tpch;
DROP DATABASE

Time: 199.566789ms

root@:26257/> create database tpch;
CREATE DATABASE

Time: 13.625286ms

root@:26257/> RESTORE tpch.* FROM 'nodelocal:/Users/rslee/data/backup/tpch';
+--------------------+-----------+--------------------+---------+---------------+----------------+------------+
|       job_id       |  status   | fraction_completed |  rows   | index_entries | system_records |   bytes    |
+--------------------+-----------+--------------------+---------+---------------+----------------+------------+
| 273132649904996352 | succeeded |                  1 | 8661245 |      56229775 |              0 | 1866390204 |
+--------------------+-----------+--------------------+---------+---------------+----------------+------------+
(1 row)

Time: 1m1.613409236s

root@:26257/> drop database tpch;
DROP DATABASE

Time: 239.433046ms

root@:26257/> create database tpch;
CREATE DATABASE

Time: 7.356251ms

root@:26257/> RESTORE tpch.* FROM 'nodelocal:/Users/rslee/data/backup/tpch';

- What did you expect to see?

- What did you see instead?

driver: bad connection
connection lost; opening new connection: all session settings will be lost
error retrieving the transaction status: dial tcp [::1]:26257: getsockopt: connection refused
connection lost; opening new connection: all session settings will be lost
error retrieving the database name: dial tcp [::1]:26257: getsockopt: connection refused
root@:26257 ?> 2017/08/22 13:46:05 Process 56415 exited with status 255

[cockroach.roberts-MBP.rslee.2017-08-22T13_47_10-04_00.056516.log.gz](https://github.com/cockroachdb/cockroach/files/1242888/cockroach.roberts-MBP.rslee.2017-08-22T13_47_10-04_00.056516.log.gz)

**Comment by benesch** [comment]
```panic: IO error: /Users/rslee/bin/cockroach-data/1/002493.sst: Too many open files in system```

**Comment by benesch** [comment]
which was quickly followed up by a panic in our error reporting. At least two bugs here.

**Comment by dt** [comment]
I think @tschottdorf has fixed or is fixing the settings refactor NPE in crash reporting.

**Comment by tschottdorf** [comment]
Thought I had fixed it. If it still happens on `master`, my help is only a stacktrace away.

**Comment by dt** [comment]
@tschottdorf attached log indicated version v1.1-alpha.20170817, has stack trace:
```
github.com/cockroachdb/cockroach/pkg/util/log.sendCrashReport(0x6cac8e8, 0xc420b54090, 0x0, 0x0, 0x5467480, 0xc4202160f0, 0x4)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:150 +0x37
github.com/cockroachdb/cockroach/pkg/util/log.ReportPanic(0x6cac8e8, 0xc420b54090, 0x0, 0x0, 0x5520e60, 0xc420368040, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:100 +0x320
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc420646120, 0x6cac8e8, 0xc420b54090)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:179 +0xab
panic(0x5520e60, 0xc420368040)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/storage.addSSTablePreApply(0x6cac8e8, 0xc420b17050, 0xc420476000, 0x61ef900, 0xc42049c140, 0x61e79a0, 0xc4203f7ec0, 0x8, 0x16, 0xc4201f0db0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:612 +0x494
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).processRaftCommand(0xc420b11500, 0x6cac8e8, 0xc420b17050, 0xc4201f0d90, 0x8, 0x8, 0x16, 0x200000002, 0x2, 0x14dd3c30cf1c4613, ...)
...
```

**Comment by tschottdorf** [comment]
Yeah, this isn't in an alpha yet. The commit is 5 days old.

**Comment by benesch** [comment]
Fixed by #17991.
